okteto/okteto
# DEV-758: Implement ssh-forwarder to contact with ssh-agent running in the cluster when executing remote operations (deploy, destroy, test)

# Proposed changes

Fixes DEV-758

Implement ssh-forwarder process to use ssh-agent running in the Okteto cluster when some command defined in an Okteto Manifest needs ssh, and that Okteto Manifest is executed remotely (okteto deploy --remote, okteto destroy --remote, okteto test).

Main changes included in the PR are:
* As part of any subcommand of `remote-run` command, start a go routine which listens in a local UNIX socket and forward everything received there to the ssh-agent running in the Okteto cluster.
* Modify the Dockerfile generated for remote operations (deploy, destroy and test) to define `SSH_AUTH_SOCK` env var if the ssh-agent host name is defined
* Don't mount anymore the local agent running in the machine
* Retrieve from the cluster metadata query the information related to the ssh-agent (hostname, internal IP and port), and pass it to the `remote-run` command through env variables defined in the generated Dockerfile
* Mount always the known_hosts file if exists locally

## How to validate


1. Update your dev environment with the latest changes of master
2. Build the image from this branch using your environment context.
  * `okteto ctx use https://...`
  * `okteto build -f Dockerfile -t okteto.dev/cli:test-pr` 
5. Take the expanded URL of the image build, and define an admin variable named `OKTETO_REMOTE_CLI_IMAGE` with the URL as value
6. Build the go binary from this branch
7. Take the public SSH of your cluster and add it to your user in GitHub (https://www.okteto.com/docs/admin/private-repositories/ssh-key/)
8. Create an Okteto Manifest like the following one:

```
deploy:
  remote: true
  commands:
    - ssh-add -L
    - git clone <private-repo>
```
9. Execute `okteto deploy` with the binary created from this branch. You should see how it lists one identity, and it should match with the SSH key showed in the admin console of your environment. You should also be able to clone the private repository.
10. You can also try with `destroy` and `test` sections of the Okteto Manifest, as they should work too

## CLI Quality Reminders 🔧

For both authors and reviewers:

- Scrutinize for potential regressions
- Ensure key automated tests are in place
- Build the CLI and test using the validation steps
- Assess Developer Experience impact (log messages, performances, etc)
- If too broad, consider breaking into smaller PRs
- Adhere to our [code style](https://github.com/okteto/okteto/blob/master/docs/code-style.md) and [code review](https://github.com/okteto/okteto/blob/master/docs/code-review.md) guidelines
